src/main/scala/dsptools/numbers/chisel_types/IntervalTypeClass.scala

```diff
@@ -0,0 +1,244 @@
+  def isWhole(a: Interval): Bool = a === floor(a)
+  // Truncate = round towards zero (integer part without fractional bits)
+  def truncate(a: Interval): Interval = {
+    Mux(isSignNegative(a),
+      ceil(a),
+      floor(a)
+    )
+  }
+  // ceil, round moved to IntervalReal to get access to ring
+}
+
+trait ConvertableToInterval extends ConvertableTo[Interval] with hasContext {
+  def fromShort(n: Short): Interval = fromInt(n.toInt)
+  def fromByte(n: Byte): Interval = fromInt(n.toInt)
+  def fromInt(n: Int): Interval = fromBigInt(BigInt(n))
+  def fromFloat(n: Float): Interval = fromDouble(n.toDouble)
+  def fromBigDecimal(n: BigDecimal): Interval = fromDouble(n.doubleValue)
+  def fromLong(n: Long): Interval = fromBigInt(BigInt(n))
+  def fromType[B](n: B)(implicit c: ConvertableFrom[B]): Interval = fromDouble(c.toDouble(n))
+  def fromBigInt(n: BigInt): Interval = n.doubleValue.I(0.BP)
+  // If no binary point is specified, use the default one provided by DspContext
+  // TODO: Should you instead be specifying a max width so you can get the most resolution for a given width?
+  def fromDouble(n: Double): Interval = n.I(context.binaryPoint.getOrElse(0).BP)
+  override def fromDouble(d: Double, a: Interval): Interval = {
+    require(a.range.binaryPoint.known, "Binary point must be known!")
+    d.I(a.range.binaryPoint)
+  }
+  override def fromDoubleWithFixedWidth(d: Double, a: Interval): Interval = {
+    require(a.range.binaryPoint.known, "Binary point must be known!")
+    require(a.widthKnown, "Interval width not known!")
+    val sintBits = BigInt(d.toInt).bitLength + 1
+    require(sintBits + a.range.binaryPoint.get <= a.getWidth, "Lit can't fit in prototype Interval bitwidth")
+    Interval.fromDouble(value = d, width = a.getWidth, binaryPoint = a.range.binaryPoint.get)
+  }
+}
+
+trait ConvertableFromInterval extends ChiselConvertableFrom[Interval] with hasContext {
+  // intPart depends on truncate
+  // asReal depends on shifting fractional bits up
+  override def asFixed(a: Interval): FixedPoint = a.asFixedPoint(a.binaryPoint)
+  // Note: need to properly align binary points
+  def asFixed(a: Interval, proto: FixedPoint): FixedPoint = ???
+  def toInterval(a: Interval, proto: Interval): Interval = a
+  override def toInterval(a: Interval): Interval = a
+}
+
+trait BinaryRepresentationInterval extends BinaryRepresentation[Interval] with hasContext {
+  def shl(a: Interval, n: Int): Interval = a << n
+  def shl(a: Interval, n: UInt): Interval = a << n
+
+  // Note: This rounds to negative infinity (smallest abs. value for negative #'s is -LSB)
+  def shr(a: Interval, n: Int): Interval = a >> n
+  def shr(a: Interval, n: UInt): Interval = a >> n
+
+  // mul2 consistent with shl
+  // signBit relies on Signed
+
+  // Retains significant digits while dividing
+  override def div2(a: Interval, n: Int): Interval = {
+    require(a.binaryPoint.known, "div2 Binary point must be known!")
+    val outFull = a.shiftLeftBinaryPoint(n) >> n
```

Comment:
Maybe I'm just confused- is the `>>` not removing significant digits here?

Reply:
Yeah uhh that's confusing -- just following Adam's firrtl convention. >> removes significant digits, so you first have to LSB pad  binary points with slbp first :\